design/cookie-rewrite-design.md

```diff
@@ -0,0 +1,196 @@
+
+Contour's ["cookie" load balancing strategy](https://projectcontour.io/docs/v1.18.1/config/request-routing/#session-affinity) enables users to implement "session affinity" by configuring Envoy to generate a cookie that can be returned in subsequent requests to route to a consistent backend app instance.
+A commonly requested feature for Contour is to make attributes of this cookie configurable.
+Users may want to apply security or other settings to ensure browsers treat these cookies appropriately.
+For example, the `SameSite` and `Secure` attributes are currently not set by Envoy when it generates the `X-Contour-Session-Affinity`.
+
+In addition, some users may want to ensure certain attributes are set on application generated cookies.
+These attributes may be things an application may not be able to accurately set, without prior knowledge of how the application is deployed.
+For example, if Contour is in use to rewrite the path or hostname of a request before it reaches an application backend, the application may not be able to accurately set the `Path` and `Domain` attributes in a `Set-Cookie` response header.
+
+Envoy does not at the moment provide a native mechanism with which HTTP cookies can be modified.
+Envoy does of course provide the ability to write Lua code to implement custom logic for processing HTTP requests and responses.
+It appears most community solutions to cookie rewriting utilize this method to customize HTTP cookies.
+However, the Envoy project does appear to be open to an API change to allow some form of [cookie rewriting](https://github.com/envoyproxy/envoy/issues/15612).
+
+## Goals
+- Provide users ability to follow best practices in cookie attribute settings
+- Offer features in Contour/Envoy provided by other L7 proxies
+  - [Nginx](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cookie_domain)
+  - [Apache HTTPD](http://httpd.apache.org/docs/2.2/mod/mod_proxy.html#proxypassreversecookiedomain)
+- Address various open issues related to cookie attributes
+  - https://github.com/projectcontour/contour/issues/2965
+  - https://github.com/projectcontour/contour/issues/3142
+
+## Non Goals
+- Provide full ability to configure Contour/Envoy with arbitrary Lua code
+- Allow users to customize all aspects of an HTTP cookie (value, other attributes not mentioned below, etc.)
+
+## High-Level Design
+`HTTPProxy` will have new fields to configure cookie rewrite policy.
+Cookie attributes related to infrastructure and security considerations will be customizable.
+Other fields, cookie names, and values will not be made customizable.
+The fields we will offer as customizable are as follows:
+- Path
+- Domain
+- Secure
+- SameSite
+
+Initially, the added field for cookie rewriting will be added to the `Route` and `Service` elements of the `HTTPProxy`.
```

Comment:
What's the use-case of having the `Service` element be able to customize this? Feels like a route type configuration, but curious if there's a use-case I'm not thinking of.

Reply:
I just added it here to match the header request/response policies, but for all of these could be that one of the services (but not all) you forward traffic to returns some header/cookie values you want to modify. having it on the route only might mean you can't have customization per backend service, which is why i assume we added the per-service header customization options?

Reply:
gotcha, that makes sense!